File: user_scripts/src/components/user_scripts/android/java/src/org/chromium/components/user_scripts/UserScriptsBridge.java
/*
    This file is part of Bromite.

    Bromite is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Bromite is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Bromite. If not, see <https://www.gnu.org/licenses/>.
*/

package org.chromium.components.user_scripts;

import java.util.ArrayList;
import java.util.List;
import java.lang.ref.WeakReference;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import androidx.annotation.Nullable;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.base.Log;
import org.chromium.ui.base.WindowAndroid;

import org.chromium.components.user_scripts.ScriptListPreference;
import org.chromium.components.user_scripts.IUserScriptsUtils;

@JNINamespace("user_scripts")
public class UserScriptsBridge {
    static WeakReference<ScriptListPreference> observer;

    private static IUserScriptsUtils utilInstance;

    public static void registerUtils(IUserScriptsUtils instance) {
        utilInstance = instance;
    }

    public static IUserScriptsUtils getUtils() {
        return utilInstance;
    }

    public static boolean isFeatureEnabled() {
        return UserScriptsBridgeJni.get().isFeatureEnabled();
    }

    public static boolean isUserEnabled() {
        return UserScriptsBridgeJni.get().isUserEnabled();
    }

    public static void setUserEnabled(boolean enabled) {
        UserScriptsBridgeJni.get().setUserEnabled(enabled);
    }

    public static void RemoveScript(String key) {
        UserScriptsBridgeJni.get().removeScript(key);
    }

    public static void SetScriptEnabled(String key,
                                        boolean enabled) {
        UserScriptsBridgeJni.get().setScriptEnabled(key, enabled);
    }

    public static void Reload() {
        UserScriptsBridgeJni.get().reload();
    }

    public static void SelectAndAddScriptFromFile(WindowAndroid window) {
        Context context = window.getContext().get();

        Intent fileSelector = new Intent(Intent.ACTION_OPEN_DOCUMENT);
        fileSelector.addCategory(Intent.CATEGORY_OPENABLE);
        fileSelector.setType("*/*");
        fileSelector.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);

        window.showIntent(fileSelector,
            new WindowAndroid.IntentCallback() {
                @Override
                public void onIntentCompleted(WindowAndroid window, int resultCode, Intent data) {
                    if (data == null) return;
                    Uri filePath = data.getData();
                    UserScriptsBridgeJni.get().tryToInstall(filePath.toString());
                }
            },
            null);
    }

    public static void TryToInstall(String ScriptFullPath) {
        UserScriptsBridgeJni.get().tryToInstall(ScriptFullPath);
    }

    public static List<ScriptInfo> getUserScriptItems() {
        List<ScriptInfo> list = new ArrayList<>();
        try {
            String json = UserScriptsBridgeJni.get().getScriptsInfo();

            JSONObject jsonObject = new JSONObject(json);

            JSONArray scripts = jsonObject.names();
            if (scripts != null) {
                Log.i("User Scripts Loaded", json);
                Log.i("Totals scripts", Integer.toString(scripts.length()));
                for (int i = 0; i < scripts.length(); i++) {
                    String key = (String) scripts.get(i);
                    JSONObject script = jsonObject.getJSONObject(key);

                    ScriptInfo si = new ScriptInfo();
                    si.Key = key;
                    list.add(si);

                    if(script.has("name")) si.Name = script.getString("name");
                    if(script.has("description")) si.Description = script.getString("description");
                    if(script.has("version")) si.Version = script.getString("version");
                    if(script.has("file_path")) si.FilePath = script.getString("file_path");
                    if(script.has("url_source")) si.UrlSource = script.getString("url_source");
                    si.Enabled = script.getBoolean("enabled");
                }
            }
        } catch (Exception e) {
            Log.e("User Scripts Load Error", e.toString());
        }
        return list;
    }

    public static void RegisterLoadCallback(ScriptListPreference caller) {
        UserScriptsBridgeJni.get().registerLoadCallback();
        observer = new WeakReference<ScriptListPreference>(caller);
    }

    @CalledByNative
    private static void shouldRefreshUserScriptList() {
        ScriptListPreference reference = observer.get();
        if (reference != null) {
            reference.NotifyScriptsChanged();
        }
    }

    @CalledByNative
    private static void onUserScriptLoaded(boolean result, String error) {
        ScriptListPreference reference = observer.get();
        if (reference != null) {
            reference.OnUserScriptLoaded(result, error);
        }
    }

    @NativeMethods
    interface Natives {
        boolean isFeatureEnabled();

        boolean isUserEnabled();
        void setUserEnabled(boolean enabled);

        String getScriptsInfo();

        void removeScript(String scriptKey);
        void setScriptEnabled(String scriptKey, boolean enabled);

        void reload();
        void selectAndAddScriptFromFile(WindowAndroid window);
        void tryToInstall(String scriptFullPath);

        void registerLoadCallback();
    }

}

File: user_scripts/src/components/user_scripts/android/user_scripts_bridge.cc
/*
    This file is part of Bromite.

    Bromite is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Bromite is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Bromite. If not, see <https://www.gnu.org/licenses/>.
*/

#include <jni.h>
#include <algorithm>
#include <string>
#include <vector>
#include <sstream>
#include <iterator>

#include "base/android/callback_android.h"
#include "base/android/jni_android.h"
#include "base/android/jni_array.h"
#include "base/android/jni_string.h"
#include "base/android/scoped_java_ref.h"
#include "components/embedder_support/android/browser_context/browser_context_handle.h"
#include "ui/android/window_android.h"

#include "components/user_scripts/android/user_scripts_jni_headers/UserScriptsBridge_jni.h"
#include "../browser/userscripts_browser_client.h"
#include "user_scripts_bridge.h"

using base::android::AttachCurrentThread;
using base::android::ConvertJavaStringToUTF8;
using base::android::ConvertUTF16ToJavaString;
using base::android::ConvertUTF8ToJavaString;
using base::android::JavaParamRef;
using base::android::JavaRef;
using base::android::ScopedJavaGlobalRef;
using base::android::ScopedJavaLocalRef;
using content::BrowserContext;

namespace {

user_scripts::UserScriptsBrowserClient* GetUserScriptsBrowserClient() {
  return user_scripts::UserScriptsBrowserClient::GetInstance();
}

class CallbackObserver : public user_scripts::UserScriptLoader::Observer {
  private:
    void OnScriptsLoaded(user_scripts::UserScriptLoader* loader,
                         content::BrowserContext* browser_context) override {
        user_scripts::ShouldRefreshUserScriptList(base::android::AttachCurrentThread());
    }

    void OnUserScriptLoaded(user_scripts::UserScriptLoader* loader,
                            bool result, const std::string& error) override {
        user_scripts::OnUserScriptLoaded(base::android::AttachCurrentThread(),
          result, error);
    }

    void OnUserScriptLoaderDestroyed(user_scripts::UserScriptLoader* loader) override {}
};

CallbackObserver* g_userscripts_loader_observer = NULL;

}

namespace user_scripts {

static jboolean JNI_UserScriptsBridge_IsFeatureEnabled(
                    JNIEnv* env) {
  return GetUserScriptsBrowserClient() != NULL;
}

static jboolean JNI_UserScriptsBridge_IsUserEnabled(
                    JNIEnv* env) {
  user_scripts::UserScriptsBrowserClient* client = GetUserScriptsBrowserClient();
  if (client == NULL) return false;
  return client->GetPrefs()->IsEnabled();
}

static void JNI_UserScriptsBridge_SetUserEnabled(
                    JNIEnv* env,
                    jboolean is_enabled) {
  user_scripts::UserScriptsBrowserClient* client = GetUserScriptsBrowserClient();
  if (client == NULL) return;
  client->GetPrefs()->SetEnabled(is_enabled);
}

static base::android::ScopedJavaLocalRef<jstring> JNI_UserScriptsBridge_GetScriptsInfo(
                    JNIEnv* env) {
  user_scripts::UserScriptsBrowserClient* client = GetUserScriptsBrowserClient();
  if (client == NULL) return ConvertUTF8ToJavaString(env, {});

  std::string json = client->GetPrefs()->GetScriptsInfo();
  return ConvertUTF8ToJavaString(env, json);
}

static void JNI_UserScriptsBridge_RemoveScript(
                    JNIEnv* env,
                    const JavaParamRef<jstring>& jscript_key) {
  user_scripts::UserScriptsBrowserClient* client = GetUserScriptsBrowserClient();
  if (client == NULL) return;

  std::string script_key = base::android::ConvertJavaStringToUTF8(jscript_key);
  client->GetLoader()->RemoveScript(script_key);
}

static void JNI_UserScriptsBridge_SetScriptEnabled(
                    JNIEnv* env,
                    const JavaParamRef<jstring>& jscript_key,
                    jboolean is_enabled) {
  user_scripts::UserScriptsBrowserClient* client = GetUserScriptsBrowserClient();
  if (client == NULL) return;

  std::string script_key = base::android::ConvertJavaStringToUTF8(jscript_key);
  client->GetLoader()->SetScriptEnabled(script_key, is_enabled);
}

static void JNI_UserScriptsBridge_Reload(
                    JNIEnv* env) {
  user_scripts::UserScriptsBrowserClient* client = GetUserScriptsBrowserClient();
  if (client == NULL) return;

  client->GetLoader()->StartLoad();
  user_scripts::ShouldRefreshUserScriptList(env);
}

static void JNI_UserScriptsBridge_SelectAndAddScriptFromFile(
                    JNIEnv* env,
                    const JavaParamRef<jobject>& jwindow_android) {
  user_scripts::UserScriptsBrowserClient* client = GetUserScriptsBrowserClient();
  if (client == NULL) return;

  client->GetLoader()->SelectAndAddScriptFromFile(
                    ui::WindowAndroid::FromJavaWindowAndroid(jwindow_android));
}

static void JNI_UserScriptsBridge_TryToInstall(JNIEnv* env,
                         const JavaParamRef<jstring>& jscript_path) {
  user_scripts::UserScriptsBrowserClient* client = GetUserScriptsBrowserClient();
  if (client == NULL) return;

  std::string script_path = base::android::ConvertJavaStringToUTF8(jscript_path);
  base::FilePath path(script_path);

  client->GetLoader()->TryToInstall(path);
}

static void JNI_UserScriptsBridge_RegisterLoadCallback(
                    JNIEnv* env) {
  user_scripts::UserScriptsBrowserClient* client = GetUserScriptsBrowserClient();
  if (client == NULL) return;

  if( g_userscripts_loader_observer == NULL ) {
    g_userscripts_loader_observer = new CallbackObserver();
    client->GetLoader()->AddObserver(g_userscripts_loader_observer);
  }
}

static void ShouldRefreshUserScriptList(JNIEnv* env) {
  Java_UserScriptsBridge_shouldRefreshUserScriptList(env);
}

static void OnUserScriptLoaded(JNIEnv* env,
              bool result, const std::string& error) {
  base::android::ScopedJavaLocalRef<jstring> j_error =
    base::android::ConvertUTF8ToJavaString(env, error);

  Java_UserScriptsBridge_onUserScriptLoaded(env, result, j_error);
}

}